packages/contractkit/src/wallets/ledger-wallet.ts

```diff
@@ -0,0 +1,222 @@
+import { isHexString, normalizeAddressWith0x, trimLeading0x } from '@celo/utils/lib/address'
+import { TransportError, TransportStatusError } from '@ledgerhq/errors'
+import Ledger from '@ledgerhq/hw-app-eth'
+import { byContractAddress } from '@ledgerhq/hw-app-eth/erc20'
+import debugFactory from 'debug'
+import * as ethUtil from 'ethereumjs-util'
+import { EncodedTransaction, Tx } from 'web3-core'
+import { Address } from '../base'
+import { EIP712TypedData, generateTypedDataHash } from '../utils/sign-typed-data-utils'
+import {
+  chainIdTransformationForSigning,
+  encodeTransaction,
+  rlpEncodedTx,
+  signatureFormatter,
+} from '../utils/signing-utils'
+import { Wallet } from './wallet'
+
+export const CELO_BASE_DERIVATION_PATH = "44'/52752'/0'/0"
+const ADDRESS_QTY = 5
+
+export async function newLedgerWalletWithSetup(
+  transport: any,
+  derivationPathIndexes?: number[],
+  baseDerivationPath?: string
+): Promise<LedgerWallet> {
+  const wallet = new LedgerWallet(derivationPathIndexes, baseDerivationPath)
+  await wallet.init(transport)
+  return wallet
+}
+
+const debug = debugFactory('kit:wallet:ledger')
+
+export class LedgerWallet implements Wallet {
+  // Account addresses are hex-encoded, lower case alphabets
+  private readonly addressDerivationPath = new Map<Address, string>()
+  private addressesRetrieved = false
+  private setupFinished = false
+  private ledger: any
+
+  /**
+   * @param derivationPathIndexes number array of "address_index" for the base derivation path.
+   * Default: Array[0..9].
+   * Example: [3, 99, 53] will retrieve the derivation paths of
+   * [`${baseDerivationPath}/3`, `${baseDerivationPath}/99`, `${baseDerivationPath}/53`]
+   * @param baseDerivationPath base derivation path. Default: "44'/52752'/0'/0"
+   */
+  constructor(
+    readonly derivationPathIndexes: number[] = Array.from(Array(ADDRESS_QTY).keys()),
+    readonly baseDerivationPath: string = CELO_BASE_DERIVATION_PATH
+  ) {
+    const invalidDPs = derivationPathIndexes.some(
+      (value) => !(Number.isInteger(value) && value >= 0)
+    )
+    if (invalidDPs) {
+      throw new Error('ledger-wallet: Invalid address index')
+    }
+  }
+
+  /**
+   * @param transport Transport to connect the ledger device
+   */
+  async init(transport: any) {
+    try {
+      if (this.setupFinished) {
```

Comment:
I believe this check is insufficient to prevent race condition. If this function is called twice before this.setupFinished is set, the initialization logic will run twice. This doesn't seem to cause any misbehavior in the current implementation, but may cause bugs in the future.

Reply:
Also, fixed in other PR #3256